Track stake key registrations independently from delegation

Anviking opened the pull request:
# Issue Number

#1835 


# Overview


- [x] Keep track of stake key registrations such that we don't try to register a key twice
- [ ] TODO: model impl
- There was a integration test, which I saw passing at some point, but I've not rebased it to this PR
- Some DB model-like tests would have been nice, but no time now

# Comments

To delegate in shelley we need to know whether we need to register the
stake key, or if it already exists. Trying to register the same key
twice will cause the delegation to fail.

In shelley there are three delegation certs:
- reg
- dereg
- delegate

We need to keep track of two things:
- Delegation status
- Is the stake key registered or not? (new in this commit)

Also new in this commit:
- Stake key de-registration inserts an uptdate to the delegation status
table that says that it is not delegating.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Try to generalize mkTx

Anviking committed:
Track stake key registrations independently from delegation

To delegate in shelley we need to know whether we need to register the
stake key, or if it already exists. Trying to register the same key
twice will cause the delegation to fail.

In shelley there are three delegation certs:
- reg
- dereg
- delegate

We need to keep track of two things:
- Delegation status
- Is the stake key registered or not? (new in this commit)

Also new in this commit:
- Stake key de-registration inserts an uptdate to the delegation status
table that says that it is not delegating.

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -604,9 +608,32 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
+                            (DelegationCertificate wid sl Nothing)
+                        pure <$> repsert
+                            (StakeKeyCertificateKey wid sl)
+                            (StakeKeyCertificate wid sl False)
```
Here we insert "deregistration" into both tables

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1712,13 +1712,17 @@ joinStakePool
         guardJoin pools (walMeta ^. #delegation) pid
 
-    let action = nextJoinAction pid walMeta
+    let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
```
To preserve the `isKeyReg then` for readability (and not have to add a new type) I removed the `nextJoinAction`-helper.

Slight duplication, but I think it's ok.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -604,9 +608,32 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
+                    W.CertRegisterKey _ ->
+                        pure <$> repsert
+                            (StakeKeyCertificateKey wid sl)
+                            (StakeKeyCertificate wid sl True)
```
~~Hmmm. This isn't quite right. Stake key are registered through their own certificates. So this is what we should be tracking. In particular, the case of ITN, the key is registered by _another mean_ than the wallet backend so we never get to create a registration certificate in the first place.~~

KtorZ replied:
Okay, went through too quickly at first when I saw the `repsert` and the boolean. Looks like this would work indeed and actually play nicely with both Jörmungandr and cardano-node.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2239,6 +2244,7 @@ instance HasSeverityAnnotation WalletLog where
         MsgPaymentCoinSelectionStart{} -> Debug
         MsgPaymentCoinSelection _ -> Debug
         MsgPaymentCoinSelectionAdjusted _ -> Debug
+        MsgIsStakeKeyRegistered _ -> Info
```
What about:

- `Info` when `False`
- `Debug` otherwise.

Anviking replied:
`False` would be expected for new wallets.

`True` would be common for people will ITN rewards.

Why not just `Info` for both?

KtorZ replied:
It's true that we aren't going to see this log line often actually. Only occurs when changing delegation settings, so I guess it's fine :+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -156,6 +156,16 @@ ProtocolParameters
+StakeKeyCertificate
+    stakeKeyCertWalletId             W.WalletId     sql=wallet_id
+    stakeKeyCertSlot                 W.SlotId       sql=slot
+    stakeKeyCertIsReg                Bool           sql=is_reg
```
I'd recommend using something more explicit than a boolean here. Boolean convey very little information and they are easy to misinterpret (like I did on my first review...). I'd suggest something like `Registration | Deregistration`, stored as plain text lowercased in the database, so it's also readable when debugging / inspecting.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -604,9 +608,32 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
+                    [Desc StakeKeyCertSlot]
+
+            return $ case val of
+                Nothing -> Left $ ErrNoSuchWallet wid
```
Hmmm. No. There could be a wallet but no key. This should return negatively about the existing of a key, but not return an error. To know whether there's a wallet, this should use the same pattern as for other functions:

```hs
            selectWallet wid >>= \case
                Nothing -> pure $ Left $ ErrNoSuchWallet wid
                Just{} -> {- ... do actual stuff ... -}
```

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -156,6 +156,16 @@ ProtocolParameters
+    stakeKeyCertSlot                 W.SlotId       sql=slot
+    stakeKeyCertIsReg                Bool           sql=is_reg
+
+    Primary stakeKeyCertWalletId stakeKeyCertSlot
```
:+1:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -676,7 +676,8 @@ fromShelleyDelegationCert = \case
 
-    SL.DCertDeleg SL.RegKey{} -> Nothing
+    SL.DCertDeleg (SL.RegKey cred) ->
+        Just $ W.CertRegisterKey $ fromStakeCredential cred
```
:+1:

Anviking committed:
Fixup ErrNoSuchWallet in readIsStakeKeyRegistered

Anviking committed:
use an explicit sum-type instead of bool for stake key

Booleans are error-prone and convey very little information without
surrounding context. A sum-type is cheap and much easier to read and
debug (even in the database!).

Co-authored-by: KtorZ <matthias.benkort@iohk.io>

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -626,14 +626,15 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
+                        [Desc StakeKeyCertSlot]
+                    return $ case val of
+                        Nothing -> Right False
+                        Just (StakeKeyCertificate _ _ isReg) -> Right isReg
```
:+1:

Anviking committed:
Extract mkTx further and export it

to allow tests to call it

KtorZ committed:
implement database abstract model for state-machine testing.

Anviking committed:
Test that wallet with existing stake key can delegate

STAKE_POOLS_JOIN_05

This is a regression test for
"Cannot join pool with ITN rewards wallet on shelley testnet"

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -369,6 +373,30 @@ spec = do
+                "passphrase": "Secure Passphrase"
+                } |]
+
+        threadDelay 4000000
```
Should use `eventually` I guess

KtorZ commented on lib/shelley/test/integration/Main.hs:
```diff
@@ -321,3 +337,53 @@ minSeverityFromEnv def var = lookupEnv var >>= \case
+
+-- | Pre-register a staking key for the STAKE_POOLS_JOIN_05 test.
+preregisterStakingKeysForTests :: NetworkLayer IO t b -> IO ()
+preregisterStakingKeysForTests nl = do
```
I am not fond of having this function here but I guess it'll do for now. Ideally, I'd have put that where the rest of the "setup & fixture" work is happening (i.e. in the "Launch" module).

KtorZ commented on lib/shelley/test/integration/Main.hs:
```diff
@@ -321,3 +337,53 @@ minSeverityFromEnv def var = lookupEnv var >>= \case
+            , outputs = []
+            , change = []
+            , deposit = 100000 -- keyDeposit
+            }
```
CoinSelection is a `Monoid`, so you can write instead: 

```hs
mempty { inputs = [(txIn, dummyOut)], deposit = 100000 }
```

Also, note that the key deposit is available in protocol parameters, would be better than hard-coding it.

Anviking replied:
> key deposit is available in protocol parameters

👍 

> CoinSelection is a Monoid

Didn't think about it, but in this case I found it useful to explicitly see the new fields of `CoinSelection`.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -1844,19 +1845,30 @@ instance NFData ChimericAccount
+data StakeKeyCertificate
+    = StakeKeyRegistration
+    | StakeKeyDeregistration
+    deriving (Generic, Show, Read, Eq)
```
There are so many different kinds of certificate types in here already, for multiple purposes. I'd preferred to keep it closer to the DB.

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -369,6 +373,30 @@ spec = do
+                "passphrase": "Secure Passphrase"
+                } |]
+
+        threadDelay 4000000
```
Why the delay?

Anviking replied:
We try to join immediately when the wallet is created. It has no time to find the existing key reg cert.

Should probably use `eventually` though. Just sprinkled it in to get it passing.

KtorZ committed:
move stake key pre-registration test code with other fixtures

So that we don't get to look at so many places when trying to figure out how the test cluster is setup. It's also using the cardano-cli for consistency with other fixtures (and because we have nice logging and retrying in place for this in the module already).

KtorZ reviewed (approved):
Reviewed, partially fixed and partially extended.

KtorZ committed:
remove flaky so-called 'unit' network spec for byron & shelley

These tests are quite flaky and often failing for no reason. Plus, they are testing things at the integration level since they require a running node and they are redundant with other integration scenario that are ran as part of the integration test suite

Commit:
Regenerate nix
